src/hotspot/share/opto/node.cpp

```diff
@@ -894,33 +894,30 @@ int Node::replace_edges_in_range(Node* old, Node* neww, int start, int end) {
 
 //-------------------------disconnect_inputs-----------------------------------
 // NULL out all inputs to eliminate incoming Def-Use edges.
-// Return the number of edges between 'n' and 'this'
-int Node::disconnect_inputs(Node *n, Compile* C) {
-  int edges_to_n = 0;
-
-  uint cnt = req();
-  for( uint i = 0; i < cnt; ++i ) {
-    if( in(i) == 0 ) continue;
-    if( in(i) == n ) ++edges_to_n;
-    set_req(i, NULL);
+void Node::disconnect_inputs(Compile* C) {
+  // r: a required input, null is allowed
+  // p: a precedence, null values are all at the end
+  // |r|...|r|p|...|p|null|...|null|
+  //         |                     |
+  //         req()                 len()
+  // -----------------------------------
+  uint i = 0;
+
+  for (; i < req(); ++i) {
+    if(in(i) == nullptr) continue;
+    set_req(i, nullptr);
   }
+
   // Remove precedence edges if any exist
   // Note: Safepoints may have precedence edges, even during parsing
-  if( (req() != len()) && (in(req()) != NULL) ) {
-    uint max = len();
-    for( uint i = 0; i < max; ++i ) {
-      if( in(i) == 0 ) continue;
-      if( in(i) == n ) ++edges_to_n;
-      set_prec(i, NULL);
-    }
+  // Precedences have no embedded NULL
+  for (; i < len() && in(i) != nullptr; ++i) {
+    set_prec(i, nullptr);
   }
```

Comment:
Something like next:
```
   for (uint i = req(); i < len(); ++i) {
     if (in(i) != nullptr) {
       set_prec(i, nullptr);
     }
   }
```

Reply:
This is trickier than I thought. `set_prec(i, nullptr)` will refill _in[i] with the last non-null value, or NULL.  
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/node.cpp#L1030

It means iteration from req() to len() skip a node if there are more than one non-null precedences.

Reply:
I will fix it in this issue. 
https://bugs.openjdk.java.net/browse/JDK-8254369